// @terra-money/wallet-provider@3.11.2/StaticWalletProvider.js
import { WalletContext, WalletStatus, } from '@terra-money/use-wallet';
import React, { useMemo } from 'react';
export function StaticWalletProvider({ children, defaultNetwork, status = WalletStatus.INITIALIZING, availableConnectTypes = [], availableInstallTypes = [], availableConnections = [], availableInstallations = [], wallets = [], supportFeatures = new Set(), connection = undefined, }) {
    const state = useMemo(() => {
        return {
            availableConnectTypes,
            availableInstallTypes,
            availableConnections,
            availableInstallations,
            status,
            network: defaultNetwork,
            wallets,
            supportFeatures,
            connection,
            install: () => {
                throw new Error('not implemented!');
            },
            connect: () => {
                throw new Error('not implemented!');
            },
            connectReadonly: () => {
                throw new Error('not implemented!');
            },
            disconnect: () => {
                throw new Error('not implemented!');
            },
            post: () => {
                throw new Error('not implemented!');
            },
            sign: () => {
                throw new Error('not implemented!');
            },
            signBytes: () => {
                throw new Error('not implemented!');
            },
            refetchStates: () => {
                throw new Error('not implemented!');
            },
            recheckStatus: () => {
                throw new Error('not implemented!');
            },
            isChromeExtensionCompatibleBrowser: () => {
                throw new Error('not implemented!');
            },
            hasCW20Tokens: () => {
                throw new Error('not implemented!');
            },
            addCW20Tokens: () => {
                throw new Error('not implemented!');
            },
            hasNetwork: () => {
                throw new Error('not implemented!');
            },
            addNetwork: () => {
                throw new Error('not implemented!');
            },
        };
    }, [
        availableConnectTypes,
        availableInstallTypes,
        availableConnections,
        availableInstallations,
        status,
        defaultNetwork,
        wallets,
        supportFeatures,
        connection,
    ]);
    return (React.createElement(WalletContext.Provider, { value: state }, children));
}
