// ~proto/messages/vesting/msgCreateClawbackVestingAccount.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMsgCreateClawbackVestingAccount = void 0;
const vesting = __importStar(require("../../proto/evmos/vesting/v1/tx"));
const cosmosVesting = __importStar(require("../../proto/cosmos/vesting/v1beta1/vesting"));
const base = __importStar(require("../../proto/cosmos/base/v1beta1/coin"));
const timestamp_1 = require("../../proto/google/protobuf/timestamp");
const toProtoPeriod = ({ length, amount }) => new cosmosVesting.cosmos.vesting.v1beta1.Period({
    length,
    amount: amount.map(({ denom, amount }) => new base.cosmos.base.v1beta1.Coin({
        denom,
        amount,
    })),
});
function createMsgCreateClawbackVestingAccount(fromAddress, toAddress, startTime, lockupPeriods, vestingPeriods, merge) {
    const msg = new vesting.evmos.vesting.v1.MsgCreateClawbackVestingAccount({
        from_address: fromAddress,
        to_address: toAddress,
        start_time: new timestamp_1.google.protobuf.Timestamp({
            seconds: startTime,
        }),
        lockup_periods: lockupPeriods.map(toProtoPeriod),
        vesting_periods: vestingPeriods.map(toProtoPeriod),
        merge,
    });
    return {
        message: msg,
        path: 'evmos.vesting.v1.MsgCreateClawbackVestingAccount',
    };
}
exports.createMsgCreateClawbackVestingAccount = createMsgCreateClawbackVestingAccount;
