// @ianvs/prettier-plugin-sort-imports@3.7.1/lib/src/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.newLineNode = exports.TYPES_SPECIAL_WORD = exports.THIRD_PARTY_MODULES_SPECIAL_WORD = exports.BUILTIN_MODULES = exports.mergeableImportFlavors = exports.importFlavorIgnore = exports.importFlavorSideEffect = exports.importFlavorType = exports.importFlavorValue = exports.chunkTypeOther = exports.chunkTypeUnsortable = exports.newLineCharacters = exports.jsx = exports.typescript = exports.flow = void 0;
var types_1 = require("@babel/types");
var module_1 = require("module");
exports.flow = 'flow';
exports.typescript = 'typescript';
exports.jsx = 'jsx';
exports.newLineCharacters = '\n\n';
exports.chunkTypeUnsortable = 'unsortable';
exports.chunkTypeOther = 'other';
/** Value imports (including top-level default imports) - import {Thing} from ... or import Thing from ... */
exports.importFlavorValue = 'value';
/** import type {} from ...  */
exports.importFlavorType = 'type';
exports.importFlavorSideEffect = 'side-effect';
exports.importFlavorIgnore = 'prettier-ignore';
exports.mergeableImportFlavors = [
    exports.importFlavorValue,
    exports.importFlavorType,
];
/*
 * Used to mark the position between RegExps,
 * where the not matched imports should be placed
 */
exports.BUILTIN_MODULES = "^(?:node:)?(?:".concat(module_1.builtinModules.join('|'), ")$");
exports.THIRD_PARTY_MODULES_SPECIAL_WORD = '<THIRD_PARTY_MODULES>';
exports.TYPES_SPECIAL_WORD = '<TYPES>';
var PRETTIER_PLUGIN_SORT_IMPORTS_NEW_LINE = 'PRETTIER_PLUGIN_SORT_IMPORTS_NEW_LINE';
exports.newLineNode = (0, types_1.expressionStatement)((0, types_1.stringLiteral)(PRETTIER_PLUGIN_SORT_IMPORTS_NEW_LINE));
