// @evmos/proto@0.1.27/dist/messages/vesting/msgClawback.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMsgClawback = void 0;
const vesting = __importStar(require("../../proto/evmos/vesting/v1/tx"));
function createMsgClawback(funderAddress, accountAddress, destAddress) {
    const msg = new vesting.evmos.vesting.v1.MsgClawback({
        funder_address: funderAddress,
        account_address: accountAddress,
        dest_address: destAddress,
    });
    return {
        message: msg,
        path: 'evmos.vesting.v1.MsgClawback',
    };
}
exports.createMsgClawback = createMsgClawback;
