// @terra-money/wallet-controller@3.11.2/_commonjs/operators/getExtensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExtensions = void 0;
const rxjs_1 = require("rxjs");
const fetch_1 = require("rxjs/fetch");
const operators_1 = require("rxjs/operators");
const browser_check_1 = require("../utils/browser-check");
const FALLBACK = {
    whitelist: [
        {
            name: 'Terra Station',
            identifier: 'station',
            icon: 'https://assets.terra.money/icon/station-extension/icon.png',
            urls: [
                {
                    browser: 'chrome',
                    url: 'https://chrome.google.com/webstore/detail/terra-station/aiifbnbfobpmeekipheeijimdpnlpgpp',
                },
                {
                    browser: 'firefox',
                    url: 'https://addons.mozilla.org/en-US/firefox/addon/terra-station-wallet/',
                },
            ],
        },
        {
            name: 'XDEFI Wallet',
            identifier: 'xdefi-wallet',
            icon: 'https://xdefi-prod-common-ui.s3.eu-west-1.amazonaws.com/logo.svg',
            urls: [
                {
                    browser: 'chrome',
                    url: 'https://chrome.google.com/webstore/detail/xdefi-wallet/hmeobnfnfcmdkdcmlblgagmfpfboieaf',
                },
            ],
        },
    ],
};
function getExtensions() {
    const currentBrowser = (0, browser_check_1.getDesktopBrowserType)(navigator.userAgent);
    if (!currentBrowser) {
        return (0, rxjs_1.of)([]);
    }
    return (0, fetch_1.fromFetch)('https://assets.terra.money/extensions.json').pipe((0, rxjs_1.switchMap)((res) => {
        if (res.ok) {
            return res.json();
        }
        else {
            return (0, rxjs_1.of)(FALLBACK);
        }
    }), (0, operators_1.catchError)(() => {
        return (0, rxjs_1.of)(FALLBACK);
    }), (0, operators_1.map)(({ whitelist }) => {
        return whitelist
            .filter(({ urls }) => urls.some(({ browser }) => currentBrowser === browser))
            .map(({ name, identifier, icon, urls }) => {
            return {
                name,
                identifier,
                icon,
                url: urls.find(({ browser }) => currentBrowser === browser).url,
            };
        });
    }));
}
exports.getExtensions = getExtensions;
