// spinners-react@1.0.7/lib/cjs/style-inject.es-03cdf6cd.js
'use strict';

var withSharedProps = require('./withSharedProps-f2576f22.js');

var defaultProps = {
    speed: 100,
    still: false,
    thickness: 100,
};
var secondaryColorDefaultProps = withSharedProps.__assign(withSharedProps.__assign({}, defaultProps), { secondaryColor: 'rgba(0,0,0,0.44)' });

function styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') { return; }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

exports.defaultProps = defaultProps;
exports.secondaryColorDefaultProps = secondaryColorDefaultProps;
exports.styleInject = styleInject;
