// @terra-money/web-extension-interface@3.11.2/_commonjs/models/tx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deserializeTx = exports.serializeTx = exports.WebExtensionTxStatus = void 0;
const terra_js_1 = require("@terra-money/terra.js");
var WebExtensionTxStatus;
(function (WebExtensionTxStatus) {
    WebExtensionTxStatus["PROGRESS"] = "PROGRESS";
    WebExtensionTxStatus["SUCCEED"] = "SUCCEED";
    WebExtensionTxStatus["FAIL"] = "FAIL";
    WebExtensionTxStatus["DENIED"] = "DENIED";
})(WebExtensionTxStatus = exports.WebExtensionTxStatus || (exports.WebExtensionTxStatus = {}));
function serializeTx(tx) {
    var _a, _b, _c;
    return {
        msgs: tx.msgs.map((msg) => msg.toJSON(tx.isClassic)),
        fee: (_a = tx.fee) === null || _a === void 0 ? void 0 : _a.toJSON(tx.isClassic),
        memo: tx.memo,
        gasPrices: (_b = tx.gasPrices) === null || _b === void 0 ? void 0 : _b.toString(),
        gasAdjustment: (_c = tx.gasAdjustment) === null || _c === void 0 ? void 0 : _c.toString(),
        feeDenoms: tx.feeDenoms,
    };
}
exports.serializeTx = serializeTx;
function deserializeTx(tx) {
    const msgs = tx.msgs.map((msg) => JSON.parse(msg));
    const isProto = '@type' in msgs[0];
    return {
        ...tx,
        msgs: msgs.map((msg) => (isProto ? terra_js_1.Msg.fromData(msg, tx.isClassic) : terra_js_1.Msg.fromAmino(msg, tx.isClassic))),
        fee: tx.fee
            ? isProto
                ? terra_js_1.Fee.fromData(JSON.parse(tx.fee))
                : terra_js_1.Fee.fromAmino(JSON.parse(tx.fee))
            : undefined,
    };
}
exports.deserializeTx = deserializeTx;
