// @terra-money/wallet-types@3.11.2/wallet.js
import { AccAddress } from '@terra-money/terra.js';
import { WalletStatus, } from './types';
export function createConnectedWallet({ connection, post, sign, signBytes, supportFeatures, wallets, status, network, }) {
    try {
        if (status === WalletStatus.WALLET_CONNECTED &&
            wallets.length > 0 &&
            AccAddress.validate(wallets[0].terraAddress) &&
            !!connection) {
            const { terraAddress, connectType, design } = wallets[0];
            return {
                network,
                terraAddress: terraAddress,
                walletAddress: terraAddress,
                design,
                post: (tx) => {
                    return post(tx, terraAddress);
                },
                sign: (tx) => {
                    return sign(tx, terraAddress);
                },
                signBytes: (bytes) => {
                    return signBytes(bytes, terraAddress);
                },
                availablePost: supportFeatures.has('post'),
                availableSign: supportFeatures.has('sign'),
                availableSignBytes: supportFeatures.has('sign-bytes'),
                connectType,
                connection,
            };
        }
        else {
            return undefined;
        }
    }
    catch (_a) {
        return undefined;
    }
}
export function createInstallableWallets({ status, installations, }) {
    if (status === WalletStatus.WALLET_NOT_CONNECTED) {
        return installations;
    }
    return undefined;
}
