// @terra-money/wallet-controller@3.11.2/_commonjs/modules/walletconnect/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletConnectSignBytesUnspecifiedError = exports.WalletConnectTxUnspecifiedError = exports.WalletConnectTimeout = exports.WalletConnectTxFailed = exports.WalletConnectCreateTxFailed = exports.WalletConnectUserDenied = void 0;
class WalletConnectUserDenied extends Error {
}
exports.WalletConnectUserDenied = WalletConnectUserDenied;
class WalletConnectCreateTxFailed extends Error {
    constructor(message) {
        super(message);
        this.name = 'WalletConnectCreateTxFailed';
    }
}
exports.WalletConnectCreateTxFailed = WalletConnectCreateTxFailed;
class WalletConnectTxFailed extends Error {
    constructor(txhash, message, raw_message) {
        super(message);
        this.txhash = txhash;
        this.raw_message = raw_message;
        this.name = 'WalletConnectTxFailed';
    }
}
exports.WalletConnectTxFailed = WalletConnectTxFailed;
class WalletConnectTimeout extends Error {
    constructor(message) {
        super(message);
        this.name = 'WalletConnectTimeout';
    }
}
exports.WalletConnectTimeout = WalletConnectTimeout;
class WalletConnectTxUnspecifiedError extends Error {
    constructor(message) {
        super(message);
        this.name = 'WalletConnectTxUnspecifiedError';
    }
}
exports.WalletConnectTxUnspecifiedError = WalletConnectTxUnspecifiedError;
class WalletConnectSignBytesUnspecifiedError extends Error {
    constructor(message) {
        super(message);
        this.name = 'WalletConnectSignBytesUnspecifiedError';
    }
}
exports.WalletConnectSignBytesUnspecifiedError = WalletConnectSignBytesUnspecifiedError;
