// @terra-money/wallet-controller@3.11.2/_commonjs/modules/walletconnect/modal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerraWalletconnectQrcodeModal = void 0;
const qrcode_1 = require("qrcode");
const browser_check_1 = require("../../utils/browser-check");
const modal_style_1 = require("./modal.style");
class TerraWalletconnectQrcodeModal {
    constructor() {
        this.modalContainer = null;
        this.styleContainer = null;
        this.callback = null;
        this.setCloseCallback = (callback) => {
            this.callback = callback;
        };
        this.open = (uri, cb, _qrcodeModalOptions) => {
            var _a, _b;
            const modalContainer = document.createElement('div');
            const stylecontainer = document.createElement('style');
            const query = encodeURIComponent(`action=wallet_connect&payload=${encodeURIComponent(uri)}`);
            const schemeUri = `https://terrastation.page.link/?link=https://terra.money?${query}&apn=money.terra.station&ibi=money.terra.station&isi=1548434735`;
            const element = createModalElement({
                schemeUri,
                onClose: () => {
                    if (this.callback) {
                        this.callback();
                        this.callback = null;
                    }
                    this.close();
                },
            });
            if ((0, browser_check_1.isMobile)()) {
                window.location.href = schemeUri;
            }
            stylecontainer.textContent = modal_style_1.modalStyle;
            modalContainer.appendChild(element);
            (_a = document.querySelector('head')) === null || _a === void 0 ? void 0 : _a.appendChild(stylecontainer);
            (_b = document.querySelector('body')) === null || _b === void 0 ? void 0 : _b.appendChild(modalContainer);
            this.modalContainer = modalContainer;
            this.styleContainer = stylecontainer;
        };
        this.close = () => {
            var _a, _b;
            if (this.modalContainer) {
                (_a = this.modalContainer.parentElement) === null || _a === void 0 ? void 0 : _a.removeChild(this.modalContainer);
            }
            if (this.styleContainer) {
                (_b = this.styleContainer.parentElement) === null || _b === void 0 ? void 0 : _b.removeChild(this.styleContainer);
            }
            this.callback = null;
        };
    }
}
exports.TerraWalletconnectQrcodeModal = TerraWalletconnectQrcodeModal;
function createModalElement({ schemeUri, onClose, }) {
    const isMobile = (0, browser_check_1.isMobile)();
    const openTerraStationMobile = () => {
        window.location.href = schemeUri;
    };
    // ---------------------------------------------
    // container
    // ---------------------------------------------
    const container = document.createElement('div');
    container.setAttribute('class', 'wallet-wc-modal');
    // ---------------------------------------------
    // container > div.wallet-wc-modal--dim
    // ---------------------------------------------
    const dim = document.createElement('div');
    dim.setAttribute('class', 'wallet-wc-modal--dim');
    container.appendChild(dim);
    // ---------------------------------------------
    // container > div.wallet-wc-modal--content
    // ---------------------------------------------
    const content = document.createElement('section');
    content.setAttribute('class', 'wallet-wc-modal--content');
    content.setAttribute('data-device', isMobile ? 'mobile' : 'desktop');
    container.appendChild(content);
    // h1
    const title = document.createElement('h1');
    content.appendChild(title);
    const img = document.createElement('img');
    img.setAttribute('src', 'https://assets.terra.money/icon/wallet-provider/walletconnect.svg');
    img.setAttribute('style', 'width: 1em; margin-right: 10px; transform: scale(1.5) translateY(0.08em)');
    const span = document.createElement('span');
    span.textContent = 'Wallet Connect';
    title.appendChild(img);
    title.appendChild(span);
    // p
    const description = document.createElement('p');
    description.textContent =
        'Scan QR code with a WalletConnect-compatible wallet';
    content.appendChild(description);
    if (isMobile) {
        // button
        const button = document.createElement('button');
        button.addEventListener('click', openTerraStationMobile);
        button.textContent = 'Open Terra Station Mobile';
        content.appendChild(button);
    }
    else {
        // qrcode
        const canvas = document.createElement('canvas');
        (0, qrcode_1.toCanvas)(canvas, schemeUri, {
            width: 220,
            margin: 0,
            color: {
                dark: '#2043b5ff',
            },
        });
        content.appendChild(canvas);
    }
    // events
    dim.addEventListener('click', onClose);
    return container;
}
