// usehooks-ts@2.9.1/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useWindowSize = exports.useUpdateEffect = exports.useToggle = exports.useTimeout = exports.useTernaryDarkMode = exports.useStep = exports.useSsr = exports.useSessionStorage = exports.useScript = exports.useScreen = exports.useReadLocalStorage = exports.useOnClickOutside = exports.useMediaQuery = exports.useMap = exports.useLockedBody = exports.useLocalStorage = exports.useIsomorphicLayoutEffect = exports.useIsMounted = exports.useIsFirstRender = exports.useIsClient = exports.useInterval = exports.useIntersectionObserver = exports.useImageOnLoad = exports.useHover = exports.useFetch = exports.useEventListener = exports.useEventCallback = exports.useElementSize = exports.useEffectOnce = exports.useDocumentTitle = exports.useDebounce = exports.useDarkMode = exports.useCounter = exports.useCountdown = exports.useCopyToClipboard = exports.useClickAnyWhere = exports.useBoolean = void 0;
var useBoolean_1 = require("./useBoolean/useBoolean");
Object.defineProperty(exports, "useBoolean", { enumerable: true, get: function () { return __importDefault(useBoolean_1).default; } });
__exportStar(require("./useBoolean/useBoolean"), exports);
var useClickAnyWhere_1 = require("./useClickAnyWhere/useClickAnyWhere");
Object.defineProperty(exports, "useClickAnyWhere", { enumerable: true, get: function () { return __importDefault(useClickAnyWhere_1).default; } });
__exportStar(require("./useClickAnyWhere/useClickAnyWhere"), exports);
var useCopyToClipboard_1 = require("./useCopyToClipboard/useCopyToClipboard");
Object.defineProperty(exports, "useCopyToClipboard", { enumerable: true, get: function () { return __importDefault(useCopyToClipboard_1).default; } });
__exportStar(require("./useCopyToClipboard/useCopyToClipboard"), exports);
var useCountdown_1 = require("./useCountdown/useCountdown");
Object.defineProperty(exports, "useCountdown", { enumerable: true, get: function () { return __importDefault(useCountdown_1).default; } });
__exportStar(require("./useCountdown/useCountdown"), exports);
var useCounter_1 = require("./useCounter/useCounter");
Object.defineProperty(exports, "useCounter", { enumerable: true, get: function () { return __importDefault(useCounter_1).default; } });
__exportStar(require("./useCounter/useCounter"), exports);
var useDarkMode_1 = require("./useDarkMode/useDarkMode");
Object.defineProperty(exports, "useDarkMode", { enumerable: true, get: function () { return __importDefault(useDarkMode_1).default; } });
__exportStar(require("./useDarkMode/useDarkMode"), exports);
var useDebounce_1 = require("./useDebounce/useDebounce");
Object.defineProperty(exports, "useDebounce", { enumerable: true, get: function () { return __importDefault(useDebounce_1).default; } });
__exportStar(require("./useDebounce/useDebounce"), exports);
var useDocumentTitle_1 = require("./useDocumentTitle/useDocumentTitle");
Object.defineProperty(exports, "useDocumentTitle", { enumerable: true, get: function () { return __importDefault(useDocumentTitle_1).default; } });
__exportStar(require("./useDocumentTitle/useDocumentTitle"), exports);
var useEffectOnce_1 = require("./useEffectOnce/useEffectOnce");
Object.defineProperty(exports, "useEffectOnce", { enumerable: true, get: function () { return __importDefault(useEffectOnce_1).default; } });
__exportStar(require("./useEffectOnce/useEffectOnce"), exports);
var useElementSize_1 = require("./useElementSize/useElementSize");
Object.defineProperty(exports, "useElementSize", { enumerable: true, get: function () { return __importDefault(useElementSize_1).default; } });
__exportStar(require("./useElementSize/useElementSize"), exports);
var useEventCallback_1 = require("./useEventCallback/useEventCallback");
Object.defineProperty(exports, "useEventCallback", { enumerable: true, get: function () { return __importDefault(useEventCallback_1).default; } });
__exportStar(require("./useEventCallback/useEventCallback"), exports);
var useEventListener_1 = require("./useEventListener/useEventListener");
Object.defineProperty(exports, "useEventListener", { enumerable: true, get: function () { return __importDefault(useEventListener_1).default; } });
__exportStar(require("./useEventListener/useEventListener"), exports);
var useFetch_1 = require("./useFetch/useFetch");
Object.defineProperty(exports, "useFetch", { enumerable: true, get: function () { return __importDefault(useFetch_1).default; } });
__exportStar(require("./useFetch/useFetch"), exports);
var useHover_1 = require("./useHover/useHover");
Object.defineProperty(exports, "useHover", { enumerable: true, get: function () { return __importDefault(useHover_1).default; } });
__exportStar(require("./useHover/useHover"), exports);
var useImageOnLoad_1 = require("./useImageOnLoad/useImageOnLoad");
Object.defineProperty(exports, "useImageOnLoad", { enumerable: true, get: function () { return __importDefault(useImageOnLoad_1).default; } });
__exportStar(require("./useImageOnLoad/useImageOnLoad"), exports);
var useIntersectionObserver_1 = require("./useIntersectionObserver/useIntersectionObserver");
Object.defineProperty(exports, "useIntersectionObserver", { enumerable: true, get: function () { return __importDefault(useIntersectionObserver_1).default; } });
__exportStar(require("./useIntersectionObserver/useIntersectionObserver"), exports);
var useInterval_1 = require("./useInterval/useInterval");
Object.defineProperty(exports, "useInterval", { enumerable: true, get: function () { return __importDefault(useInterval_1).default; } });
__exportStar(require("./useInterval/useInterval"), exports);
var useIsClient_1 = require("./useIsClient/useIsClient");
Object.defineProperty(exports, "useIsClient", { enumerable: true, get: function () { return __importDefault(useIsClient_1).default; } });
__exportStar(require("./useIsClient/useIsClient"), exports);
var useIsFirstRender_1 = require("./useIsFirstRender/useIsFirstRender");
Object.defineProperty(exports, "useIsFirstRender", { enumerable: true, get: function () { return __importDefault(useIsFirstRender_1).default; } });
__exportStar(require("./useIsFirstRender/useIsFirstRender"), exports);
var useIsMounted_1 = require("./useIsMounted/useIsMounted");
Object.defineProperty(exports, "useIsMounted", { enumerable: true, get: function () { return __importDefault(useIsMounted_1).default; } });
__exportStar(require("./useIsMounted/useIsMounted"), exports);
var useIsomorphicLayoutEffect_1 = require("./useIsomorphicLayoutEffect/useIsomorphicLayoutEffect");
Object.defineProperty(exports, "useIsomorphicLayoutEffect", { enumerable: true, get: function () { return __importDefault(useIsomorphicLayoutEffect_1).default; } });
__exportStar(require("./useIsomorphicLayoutEffect/useIsomorphicLayoutEffect"), exports);
var useLocalStorage_1 = require("./useLocalStorage/useLocalStorage");
Object.defineProperty(exports, "useLocalStorage", { enumerable: true, get: function () { return __importDefault(useLocalStorage_1).default; } });
__exportStar(require("./useLocalStorage/useLocalStorage"), exports);
var useLockedBody_1 = require("./useLockedBody/useLockedBody");
Object.defineProperty(exports, "useLockedBody", { enumerable: true, get: function () { return __importDefault(useLockedBody_1).default; } });
__exportStar(require("./useLockedBody/useLockedBody"), exports);
var useMap_1 = require("./useMap/useMap");
Object.defineProperty(exports, "useMap", { enumerable: true, get: function () { return __importDefault(useMap_1).default; } });
__exportStar(require("./useMap/useMap"), exports);
var useMediaQuery_1 = require("./useMediaQuery/useMediaQuery");
Object.defineProperty(exports, "useMediaQuery", { enumerable: true, get: function () { return __importDefault(useMediaQuery_1).default; } });
__exportStar(require("./useMediaQuery/useMediaQuery"), exports);
var useOnClickOutside_1 = require("./useOnClickOutside/useOnClickOutside");
Object.defineProperty(exports, "useOnClickOutside", { enumerable: true, get: function () { return __importDefault(useOnClickOutside_1).default; } });
__exportStar(require("./useOnClickOutside/useOnClickOutside"), exports);
var useReadLocalStorage_1 = require("./useReadLocalStorage/useReadLocalStorage");
Object.defineProperty(exports, "useReadLocalStorage", { enumerable: true, get: function () { return __importDefault(useReadLocalStorage_1).default; } });
__exportStar(require("./useReadLocalStorage/useReadLocalStorage"), exports);
var useScreen_1 = require("./useScreen/useScreen");
Object.defineProperty(exports, "useScreen", { enumerable: true, get: function () { return __importDefault(useScreen_1).default; } });
__exportStar(require("./useScreen/useScreen"), exports);
var useScript_1 = require("./useScript/useScript");
Object.defineProperty(exports, "useScript", { enumerable: true, get: function () { return __importDefault(useScript_1).default; } });
__exportStar(require("./useScript/useScript"), exports);
var useSessionStorage_1 = require("./useSessionStorage/useSessionStorage");
Object.defineProperty(exports, "useSessionStorage", { enumerable: true, get: function () { return __importDefault(useSessionStorage_1).default; } });
__exportStar(require("./useSessionStorage/useSessionStorage"), exports);
var useSsr_1 = require("./useSsr/useSsr");
Object.defineProperty(exports, "useSsr", { enumerable: true, get: function () { return __importDefault(useSsr_1).default; } });
__exportStar(require("./useSsr/useSsr"), exports);
var useStep_1 = require("./useStep/useStep");
Object.defineProperty(exports, "useStep", { enumerable: true, get: function () { return __importDefault(useStep_1).default; } });
__exportStar(require("./useStep/useStep"), exports);
var useTernaryDarkMode_1 = require("./useTernaryDarkMode/useTernaryDarkMode");
Object.defineProperty(exports, "useTernaryDarkMode", { enumerable: true, get: function () { return __importDefault(useTernaryDarkMode_1).default; } });
__exportStar(require("./useTernaryDarkMode/useTernaryDarkMode"), exports);
var useTimeout_1 = require("./useTimeout/useTimeout");
Object.defineProperty(exports, "useTimeout", { enumerable: true, get: function () { return __importDefault(useTimeout_1).default; } });
__exportStar(require("./useTimeout/useTimeout"), exports);
var useToggle_1 = require("./useToggle/useToggle");
Object.defineProperty(exports, "useToggle", { enumerable: true, get: function () { return __importDefault(useToggle_1).default; } });
__exportStar(require("./useToggle/useToggle"), exports);
var useUpdateEffect_1 = require("./useUpdateEffect/useUpdateEffect");
Object.defineProperty(exports, "useUpdateEffect", { enumerable: true, get: function () { return __importDefault(useUpdateEffect_1).default; } });
__exportStar(require("./useUpdateEffect/useUpdateEffect"), exports);
var useWindowSize_1 = require("./useWindowSize/useWindowSize");
Object.defineProperty(exports, "useWindowSize", { enumerable: true, get: function () { return __importDefault(useWindowSize_1).default; } });
__exportStar(require("./useWindowSize/useWindowSize"), exports);
