// @terra-money/wallet-controller@3.11.2/getChainOptions.js
const FALLBACK_MAINNET = {
    name: 'mainnet',
    chainID: 'phoenix-1',
    lcd: 'https://phoenix-lcd.terra.dev',
};
const FALLBACK = {
    defaultNetwork: FALLBACK_MAINNET,
    walletConnectChainIds: {
        1: FALLBACK_MAINNET,
        0: {
            name: 'testnet',
            chainID: 'pisco-1',
            lcd: 'https://pisco-lcd.terra.dev',
        },
        2: {
            name: 'classic',
            chainID: 'columbus-5',
            lcd: 'https://columbus-lcd.terra.dev',
        },
    },
};
let cache;
export async function getChainOptions() {
    return fetch('https://assets.terra.money/chains.json')
        .then((res) => res.json())
        .then((data) => {
        const chains = Object.values(data);
        const walletConnectChainIds = chains.reduce((result, network) => {
            if (typeof network.walletconnectID === 'number') {
                result[network.walletconnectID] = network;
            }
            else if (!result[1] && network.name === 'mainnet') {
                result[1] = network;
            }
            else if (!result[0] && network.name === 'testnet') {
                result[0] = network;
            }
            else if (!result[2] && network.name === 'classic') {
                result[2] = network;
            }
            return result;
        }, {});
        const chainOptions = {
            defaultNetwork: walletConnectChainIds[1],
            walletConnectChainIds,
        };
        cache = chainOptions;
        return chainOptions;
    })
        .catch((error) => {
        console.error('Failed to fetch chains.json', error);
        return cache !== null && cache !== void 0 ? cache : FALLBACK;
    });
}
