// @terra-money/wallet-types@3.11.2/_commonjs/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignBytesUnspecifiedError = exports.TxUnspecifiedError = exports.TxFailed = exports.CreateTxFailed = exports.SignBytesFailed = exports.Timeout = exports.UserDenied = void 0;
/** User denied the request */
class UserDenied extends Error {
    constructor() {
        super('User Denied');
        this.toString = () => {
            return `[${this.name}]`;
        };
        this.name = 'UserDenied';
    }
}
exports.UserDenied = UserDenied;
/** the user did not complete the action during a specific time */
class Timeout extends Error {
    constructor(message) {
        super(message);
        this.toString = () => {
            return `[${this.name} message="${this.message}"]`;
        };
        this.name = 'Timeout';
    }
}
exports.Timeout = Timeout;
/** failed on signBytes() */
class SignBytesFailed extends Error {
    constructor(bytes, message) {
        super(message);
        this.bytes = bytes;
        this.toString = () => {
            return `[${this.name} message="${this.message}"]`;
        };
        this.name = 'SignBytesFailed';
    }
}
exports.SignBytesFailed = SignBytesFailed;
/** Failed to create tx (did not make a txhash) */
class CreateTxFailed extends Error {
    constructor(tx, message) {
        super(message);
        this.tx = tx;
        this.toString = () => {
            return `[${this.name} message="${this.message}"]\n${JSON.stringify(this.tx, null, 2)}`;
        };
        this.name = 'CreateTxFailed';
    }
}
exports.CreateTxFailed = CreateTxFailed;
/** Failed process the tx (maked a txhash) */
class TxFailed extends Error {
    constructor(tx, txhash, message, raw_message) {
        super(message);
        this.tx = tx;
        this.txhash = txhash;
        this.raw_message = raw_message;
        this.toString = () => {
            return `[${this.name} txhash="${this.txhash}" message="${this.message}"]\n${JSON.stringify(this.tx, null, 2)}\n${JSON.stringify(this.raw_message, null, 2)}`;
        };
        this.name = 'TxFailed';
    }
}
exports.TxFailed = TxFailed;
/** Unknown error */
class TxUnspecifiedError extends Error {
    constructor(tx, message) {
        super(message);
        this.tx = tx;
        this.toString = () => {
            return `[${this.name} message="${this.message}"]\n${JSON.stringify(this.tx, null, 2)}`;
        };
        this.name = 'TxUnspecifiedError';
    }
}
exports.TxUnspecifiedError = TxUnspecifiedError;
class SignBytesUnspecifiedError extends Error {
    constructor(bytes, message) {
        super(message);
        this.bytes = bytes;
        this.toString = () => {
            return `[${this.name} message="${this.message}"]\n${JSON.stringify(this.bytes, null, 2)}`;
        };
        this.name = 'SignBytesUnspecifiedError';
    }
}
exports.SignBytesUnspecifiedError = SignBytesUnspecifiedError;
