// @terra-money/wallet-types@3.11.2/_commonjs/wallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createInstallableWallets = exports.createConnectedWallet = void 0;
const terra_js_1 = require("@terra-money/terra.js");
const types_1 = require("./types");
function createConnectedWallet({ connection, post, sign, signBytes, supportFeatures, wallets, status, network, }) {
    try {
        if (status === types_1.WalletStatus.WALLET_CONNECTED &&
            wallets.length > 0 &&
            terra_js_1.AccAddress.validate(wallets[0].terraAddress) &&
            !!connection) {
            const { terraAddress, connectType, design } = wallets[0];
            return {
                network,
                terraAddress: terraAddress,
                walletAddress: terraAddress,
                design,
                post: (tx) => {
                    return post(tx, terraAddress);
                },
                sign: (tx) => {
                    return sign(tx, terraAddress);
                },
                signBytes: (bytes) => {
                    return signBytes(bytes, terraAddress);
                },
                availablePost: supportFeatures.has('post'),
                availableSign: supportFeatures.has('sign'),
                availableSignBytes: supportFeatures.has('sign-bytes'),
                connectType,
                connection,
            };
        }
        else {
            return undefined;
        }
    }
    catch (_a) {
        return undefined;
    }
}
exports.createConnectedWallet = createConnectedWallet;
function createInstallableWallets({ status, installations, }) {
    if (status === types_1.WalletStatus.WALLET_NOT_CONNECTED) {
        return installations;
    }
    return undefined;
}
exports.createInstallableWallets = createInstallableWallets;
