// ~eip712/messages/authz/generic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMsgRevokeGenericAuthorization = exports.MSG_REVOKE_GENERIC_AUTHORIZATION_TYPES = exports.createMsgGenericAuthorization = exports.MSG_GENERIC_AUTHORIZATION_TYPES = void 0;
exports.MSG_GENERIC_AUTHORIZATION_TYPES = {
    MsgValue: [
        { name: 'granter', type: 'string' },
        { name: 'grantee', type: 'string' },
        { name: 'grant', type: 'TypeGrant' },
    ],
    TypeGrant: [
        { name: 'authorization', type: 'TypeGrantAuthorization' },
        { name: 'expiration', type: 'string' },
    ],
    TypeGrantAuthorization: [
        { name: 'type', type: 'string' },
        { name: 'value', type: 'TypeGrantAuthorizationValue' },
    ],
    TypeGrantAuthorizationValue: [{ name: 'msg', type: 'string' }],
};
function createMsgGenericAuthorization(sender, botAddress, typeUrl, expires) {
    const date = new Date();
    date.setTime(expires * 1000);
    let time = date.toISOString();
    time = time.replace('.000Z', 'Z');
    return {
        type: 'cosmos-sdk/MsgGrant',
        value: {
            grant: {
                authorization: {
                    type: 'cosmos-sdk/GenericAuthorization',
                    value: {
                        msg: typeUrl,
                    },
                },
                expiration: time,
            },
            grantee: botAddress,
            granter: sender,
        },
    };
}
exports.createMsgGenericAuthorization = createMsgGenericAuthorization;
exports.MSG_REVOKE_GENERIC_AUTHORIZATION_TYPES = {
    MsgValue: [
        { name: 'granter', type: 'string' },
        { name: 'grantee', type: 'string' },
        { name: 'msg_type_url', type: 'string' },
    ],
};
function createMsgRevokeGenericAuthorization(sender, botAddress, typeUrl) {
    return {
        type: 'cosmos-sdk/MsgRevoke',
        value: {
            msg_type_url: typeUrl,
            grantee: botAddress,
            granter: sender,
        },
    };
}
exports.createMsgRevokeGenericAuthorization = createMsgRevokeGenericAuthorization;
