// ~eip712/encoding/decodeProtobuf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decodeProtobufSignDoc = exports.PROTO_MSG_TYPES = void 0;
const proto_signing_1 = require("@cosmjs/proto-signing");
const tx_1 = require("cosmjs-types/cosmos/tx/v1beta1/tx");
const tx_2 = require("cosmjs-types/cosmos/gov/v1beta1/tx");
const tx_3 = require("cosmjs-types/cosmos/staking/v1beta1/tx");
const utils_1 = require("./utils");
const decodeAmino_1 = require("./decodeAmino");
const base_1 = require("../messages/base");
const Long = require('long');
const { isLong } = require('long');
exports.PROTO_MSG_TYPES = {
    MSG_SEND: '/cosmos.bank.v1beta1.MsgSend',
    MSG_VOTE: '/cosmos.gov.v1beta1.MsgVote',
    MSG_DELEGATE: '/cosmos.staking.v1beta1.MsgDelegate',
};
function protobufTypeUrlToAminoType(typeUrl) {
    switch (typeUrl) {
        case exports.PROTO_MSG_TYPES.MSG_SEND:
            return decodeAmino_1.MSG_TYPES.MSG_SEND;
        case exports.PROTO_MSG_TYPES.MSG_VOTE:
            return decodeAmino_1.MSG_TYPES.MSG_VOTE;
        case exports.PROTO_MSG_TYPES.MSG_DELEGATE:
            return decodeAmino_1.MSG_TYPES.MSG_DELEGATE;
        default:
            throw new Error('Invalid Protobuf message type url received');
    }
}
function convertProtobufMsgToAminoMsg(obj) {
    if (typeof obj !== 'object') {
        return obj;
    }
    if (Array.isArray(obj)) {
        const formattedArray = [];
        obj.forEach((el) => {
            formattedArray.push(convertProtobufMsgToAminoMsg(el));
        });
        return formattedArray;
    }
    if (isLong(obj)) {
        return new Long(obj).toString();
    }
    const camelToSnakeCase = (str) => str.replace(/[A-Z]/g, (letter) => `_${letter.toLowerCase()}`);
    const formattedObj = {};
    Object.keys(obj).forEach((key) => {
        formattedObj[camelToSnakeCase(key)] = convertProtobufMsgToAminoMsg(obj[key]);
    });
    return formattedObj;
}
function eip712ProtobufRegistry() {
    const registry = new proto_signing_1.Registry();
    registry.register(exports.PROTO_MSG_TYPES.MSG_VOTE, tx_2.MsgVote);
    registry.register(exports.PROTO_MSG_TYPES.MSG_DELEGATE, tx_3.MsgDelegate);
    return registry;
}
function decodeProtobufSignDoc(bytes) {
    var _a, _b;
    const registry = eip712ProtobufRegistry();
    const signDoc = tx_1.SignDoc.decode(bytes);
    const txBody = tx_1.TxBody.decode(signDoc.bodyBytes);
    const authInfo = tx_1.AuthInfo.decode(signDoc.authInfoBytes);
    if (txBody.messages.length !== 1) {
        throw new Error(`Expected single message in Protobuf SignDoc but received ${txBody.messages.length}.`);
    }
    const msgBytes = txBody.messages[0];
    if (authInfo.signerInfos.length !== 1) {
        throw new Error(`Expected single signer in Protobuf SignDoc but received ${authInfo.signerInfos.length}.`);
    }
    const signer = authInfo.signerInfos[0];
    if (!authInfo.fee) {
        throw new Error('Expected fee object to be included in payload, got undefined');
    }
    if (authInfo.fee.amount.length !== 1) {
        throw new Error(`Expected single fee in Protobuf SignDoc but received ${(_a = authInfo.fee) === null || _a === void 0 ? void 0 : _a.amount.length}`);
    }
    const amount = authInfo.fee.amount[0];
    const accountNumber = signDoc.accountNumber.toString();
    const sequence = signer.sequence.toString();
    const { chainId } = signDoc;
    const { memo } = txBody;
    const protoMsg = registry.decode(msgBytes);
    const aminoMsg = {
        type: protobufTypeUrlToAminoType(msgBytes.typeUrl),
        value: convertProtobufMsgToAminoMsg(protoMsg),
    };
    let feePayer = (_b = authInfo.fee) === null || _b === void 0 ? void 0 : _b.payer;
    if (!feePayer || feePayer === '') {
        feePayer = (0, decodeAmino_1.getFeePayerFromMsg)(aminoMsg);
    }
    const gasLimit = authInfo.fee.gasLimit.toString();
    const fee = (0, base_1.generateFee)(amount.amount, amount.denom, gasLimit, feePayer);
    const type = (0, decodeAmino_1.eip712MessageType)(aminoMsg);
    const eip712Tx = (0, base_1.generateMessage)(accountNumber, sequence, chainId, memo, fee, aminoMsg);
    return (0, base_1.createEIP712)(type, (0, utils_1.parseChainId)(chainId), eip712Tx);
}
exports.decodeProtobufSignDoc = decodeProtobufSignDoc;
